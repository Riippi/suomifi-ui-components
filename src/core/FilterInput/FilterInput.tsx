import React, { Component, ChangeEvent, FocusEvent } from 'react';
import { default as styled } from 'styled-components';
import { withSuomifiDefaultProps } from '../theme/utils';
import { HtmlInput, HtmlInputProps, HtmlDiv, HtmlDivProps } from '../../reset';
import { TokensProp, InternalTokensProp } from '../theme';
import { baseStyles } from './FilterInput.baseStyles';
import { LabelText, LabelMode } from '../Form/LabelText/LabelText';
import classnames from 'classnames';
import { Omit } from '../../utils/typescript';

const baseClassName = 'fi-filter-input';

export const filterInputClassNames = {
  baseClassName,
  fullWidth: `${baseClassName}--full-width`,
  disabled: `${baseClassName}--disabled`,
  inputElementContainer: `${baseClassName}_input-element-container`,
  inputElement: `${baseClassName}_input`,
};

export interface FilterInputProps
  extends Omit<HtmlInputProps, 'type'>,
    TokensProp {
  /** FilterInput container div class name for custom styling. */
  className?: string;
  /** FilterInput container div props */
  inputContainerProps?: Omit<HtmlDivProps, 'className'>;
  /** Disable input usage */
  disabled?: boolean;
  /** Event handler to execute when clicked */
  onClick?: () => void;
  /** To execute on input text change */
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
  /** To execute on input text onBlur */
  onBlur?: (event: FocusEvent<HTMLInputElement>) => void;
  /** Label */
  labelText: string;
  /** Hide or show label. Label element is always present, but can be visually hidden.
   * @default visible
   */
  labelMode?: LabelMode;
  /** Placeholder text for input. Use only as visual aid, not for instructions. */
  visualPlaceholder?: string;
  /** FilterInput name */
  name?: string;
  /** Set components width to 100% */
  fullWidth?: boolean;
}

class BaseFilterInput extends Component<FilterInputProps> {
  render() {
    const {
      className,
      labelText,
      labelMode,
      inputContainerProps,
      visualPlaceholder,
      id,
      fullWidth,
      'aria-describedby': ariaDescribedBy,
      ...passProps
    } = this.props;

    return (
      <HtmlDiv
        {...inputContainerProps}
        className={classnames(baseClassName, className, {
          [filterInputClassNames.disabled]: !!passProps.disabled,
          [filterInputClassNames.fullWidth]: fullWidth,
        })}
      >
        <LabelText htmlFor={id} labelMode={labelMode} as="label">
          {labelText}
        </LabelText>
        <HtmlDiv className={filterInputClassNames.inputElementContainer}>
          <HtmlInput
            {...passProps}
            id={id}
            className={filterInputClassNames.inputElement}
            type="text"
            placeholder={visualPlaceholder}
            aria-describedby={ariaDescribedBy}
          />
        </HtmlDiv>
      </HtmlDiv>
    );
  }
}

const StyledFilterInput = styled(
  ({ tokens, ...passProps }: FilterInputProps & InternalTokensProp) => {
    return <BaseFilterInput {...passProps} />;
  },
)`
  ${(props) => baseStyles(props)}
`;

/**
 * <i class="semantics" />
 * Use for filtering.
 * Props other than specified explicitly are passed on to underlying input element.
 */
export class FilterInput extends Component<FilterInputProps> {
  render() {
    return <StyledFilterInput {...withSuomifiDefaultProps(this.props)} />;
  }
}
