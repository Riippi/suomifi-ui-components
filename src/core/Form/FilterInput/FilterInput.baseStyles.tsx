import { css } from 'styled-components';
import { withSuomifiTheme, TokensAndTheme } from '../../theme';
import { input, containerIEFocus, font } from '../../theme/reset';
import { absoluteFocus } from '../../theme/utils';

export const baseStyles = withSuomifiTheme(
  ({ theme }: TokensAndTheme) => css`
    & .fi-filter-input {
      ${font({ theme })('bodyText')}
      display: inline-block;
      width: 100%;
    }

    & .fi-filter-input_input-element-container {
      ${containerIEFocus({ theme })}

      &:focus-within {
        position: relative;

        &::after {
          ${absoluteFocus}
        }
      }
    }

    & .fi-filter-input_input {
      ${input({ theme })}
      background-color: ${theme.colors.whiteBase};
      min-width: 40px;
      width: 100%;
      min-height: 40px;
      padding-left: ${theme.spacing.insetL};
      border-color: ${theme.colors.depthDark3};

      &::placeholder {
        font-style: italic;
        color: ${theme.colors.depthDark2};
        opacity: 1;
      }
    }

    &.fi-filter-input--error {
      & .fi-filter-input_input {
        border-color: ${theme.colors.alertBase};
        border-width: 2px;
      }
    }

    &.fi-filter-input--disabled {
      & .fi-filter-input_input {
        color: ${theme.colors.depthBase};
        background-color: ${theme.colors.depthLight3};
        border-color: ${theme.colors.depthLight1};
      }
    }
  `,
);
