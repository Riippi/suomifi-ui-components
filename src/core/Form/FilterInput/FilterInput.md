```js
import { FilterInput } from 'suomifi-ui-components';

const tools = [
  { name: 'Jackhammer', price: 230, tax: false },
  { name: 'Hammer', price: 15, tax: true },
  { name: 'Sledgehammer', price: 36, tax: false },
  { name: 'Spade', price: 50, tax: true },
  { name: 'Powersaw', price: 150, tax: false }
];
const [filteredItems, setFilteredItems] = React.useState(tools);

// to do the magic on the component's side
const filter = (tool, query) => {
  return tool.name.includes(query);
};

<>
  <FilterInput
    items={tools}
    onFiltering={(filtered) => setFilteredItems(filtered)}
    filterRule={filter}
  />

  <div>
    <ul>
      {filteredItems.map((item, index) => {
        if (item) {
          return <li>{item.name}</li>;
        } else {
          return { item };
        }
      })}
    </ul>
  </div>
</>;
```
