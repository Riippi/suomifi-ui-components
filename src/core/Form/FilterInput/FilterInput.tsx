import React, { Component, ChangeEvent } from 'react';
import { default as styled } from 'styled-components';
import { withSuomifiDefaultProps } from '../../theme/utils';
import {
  HtmlInput,
  HtmlInputProps,
  HtmlDiv,
  HtmlDivProps,
} from '../../../reset';
import { TokensProp, InternalTokensProp } from '../../theme';
import { baseStyles } from './FilterInput.baseStyles';
import classnames from 'classnames';
import { Omit } from '../../../utils/typescript';
import { LabelText, LabelMode } from '../LabelText/LabelText';
import { AutoId } from '../../../utils/AutoId';
import { InputStatus } from '../types';
import { StatusText } from '../StatusText/StatusText';

const baseClassName = 'fi-filter-input';
const filterInputClassNames = {
  error: `${baseClassName}--error`,
  disabled: `${baseClassName}--disabled`,
  inputElementContainer: `${baseClassName}_input-element-container`,
  inputElement: `${baseClassName}_input`,
};

type FilterInputStatus = Exclude<InputStatus, 'success'>;

export interface FilterInputProps<T>
  extends Omit<HtmlInputProps, 'type'>,
    TokensProp {
  /** FilterInput container div class name for custom styling. */
  className?: string;
  /** FilterInput container div props */
  inputContainerProps?: Omit<HtmlDivProps, 'className'>;
  /** Disable input usage */
  disabled?: boolean;
  /** Placeholder text for input. Use only as visual aid, not for instructions. */
  visualPlaceholder?: string;
  /** Label */
  labelText: string;
  /** Hide or show label. Label element is always present, but can be visually hidden.
   * @default visible
   */
  labelMode?: LabelMode;
  /** Text to mark a field optional. Will be wrapped in parentheses and shown after labelText. */
  optionalText?: string;
  /**
   * 'default' | 'error'
   * @default default
   */
  status?: FilterInputStatus;
  /** Status text to be shown below the component and hint text. Use e.g. for validation error */
  statusText?: string;
  /** FilterInput name */
  name?: string;
  /** Items to be filtered */
  items: Array<T>;
  /** Returns the filtered items */
  onFiltering: (filteredItems: Array<T>) => void;
  /** Filtering rule to be used */
  filterRule: (item: T, query: string) => boolean;
}

class BaseFilterInput<T> extends Component<FilterInputProps<T>> {
  render() {
    const {
      className,
      inputContainerProps,
      visualPlaceholder,
      labelText,
      labelMode,
      optionalText,
      status,
      statusText,
      id,
      'aria-describedby': ariaDescribedBy,
      items: propItems,
      onFiltering: propOnFiltering,
      filterRule: propFilterRule,
      ...passProps
    } = this.props;

    const onChangeHandler = (event: ChangeEvent<HTMLInputElement>) => {
      const { items, onFiltering, filterRule } = this.props;
      const { value } = event.target;

      const filteredItems: T[] = items.reduce((filtered: T[], item: T) => {
        if (filterRule(item, value)) {
          filtered.push(item);
        }
        return filtered;
      }, []);

      onFiltering(filteredItems);
    };

    const statusTextId = statusText ? `${id}-statusText` : undefined;

    return (
      <HtmlDiv
        {...inputContainerProps}
        className={classnames(baseClassName, className, {
          [filterInputClassNames.disabled]: !!passProps.disabled,
          [filterInputClassNames.error]: status === 'error',
        })}
      >
        <LabelText
          htmlFor={id}
          as="label"
          labelMode={labelMode}
          optionalText={optionalText}
        >
          {labelText}
        </LabelText>
        <HtmlDiv className={filterInputClassNames.inputElementContainer}>
          <HtmlInput
            {...passProps}
            id={id}
            className={filterInputClassNames.inputElement}
            type="text"
            placeholder={visualPlaceholder}
            aria-describedby={ariaDescribedBy}
            autoComplete="off"
            aria-autocomplete="list"
            auto-capitalize="false"
            spellCheck="false"
            onChange={onChangeHandler}
          />
        </HtmlDiv>
        <StatusText id={statusTextId} status={status}>
          {statusText}
        </StatusText>
      </HtmlDiv>
    );
  }
}

const FilterInputWithoutTokens: <T>(
  props: FilterInputProps<T> & InternalTokensProp,
) => JSX.Element = ({
  // eslint-disable-next-line react/prop-types
  tokens,
  // eslint-disable-next-line react/prop-types
  id: propId,
  ...passProps
}) => (
  <AutoId id={propId}>
    {(id) => <BaseFilterInput id={id} {...passProps} />}
  </AutoId>
);

const StyledFilterInput = styled(FilterInputWithoutTokens)`
  ${(props) => baseStyles(props)}
`;

/**
 * <i class="semantics" />
 * Use for filtering.
 * Props other than specified explicitly are passed on to underlying input element.
 */
export class FilterInput<T> extends Component<FilterInputProps<T>> {
  render() {
    return <StyledFilterInput {...withSuomifiDefaultProps(this.props)} />;
  }
}
