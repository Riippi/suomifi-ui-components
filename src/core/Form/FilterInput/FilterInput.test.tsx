import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../../utils/test/axe';

import { FilterInput } from './FilterInput';

interface Product {
  name: string;
  price: number;
  tax: boolean;
}

const tools: Product[] = [
  { name: 'Jackhammer', price: 230, tax: false },
  { name: 'Hammer', price: 15, tax: true },
  { name: 'Sledgehammer', price: 36, tax: false },
  { name: 'Spade', price: 50, tax: true },
  { name: 'Powersaw', price: 150, tax: false },
];

const filter = (tool: Product, query: string) => tool.name.includes(query);

const TestFilterInput = (
  <FilterInput
    labelText="Label"
    items={tools}
    onFiltering={(filtered) => console.log(filtered)}
    filterRule={filter}
  />
);

test('should not have basic accessibility issues', async () =>
  axeTest(TestFilterInput));

test('snapshot matches', () => {
  const { container } = render(TestFilterInput);
  expect(container.firstChild).toMatchSnapshot();
});

describe('props', () => {
  describe('className', () => {
    it('has the given custom className', () => {
      const { container } = render(
        <FilterInput
          className="custom-style"
          labelText="Label"
          items={tools}
          onFiltering={(filtered) => console.log(filtered)}
          filterRule={filter}
        />,
      );
      expect(container.firstChild).toHaveClass('custom-style');
    });
  });

  describe('statusText', () => {
    it('has the status text element', () => {
      const { getByText } = render(
        <FilterInput
          statusText="Example status text"
          labelText="Label"
          items={tools}
          onFiltering={(filtered) => console.log(filtered)}
          filterRule={filter}
        />,
      );
      const statusText = getByText('Example status text');
      expect(statusText).toHaveClass('fi-status-text');
    });

    it('will be added to input aria-describedby', () => {
      const { getByRole } = render(
        <FilterInput
          statusText="Example status text"
          id="123"
          labelText="Label"
          items={tools}
          onFiltering={(filtered) => console.log(filtered)}
          filterRule={filter}
        />,
      );
      expect(getByRole('textbox')).toHaveAttribute(
        'aria-describedby',
        '123-statusText',
      );
    });
  });
});
