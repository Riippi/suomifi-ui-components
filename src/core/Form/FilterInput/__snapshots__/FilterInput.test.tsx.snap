// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot matches 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
}

.c4::-webkit-input-placeholder {
  color: inherit;
  opacity: 0.54;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2.fi-label-text .fi-label-text_label-span {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  display: block;
  margin-bottom: 10px;
  color: hsl(0,0%,16%);
}

.c2.fi-label-text .fi-label-text_label-span .fi-label-text_optionalText {
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c1 .fi-filter-input {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  display: inline-block;
  width: 100%;
}

.c1 .fi-filter-input_input-element-container > input:focus {
  outline-color: hsl(196,77%,44%);
  outline-width: 2px;
  outline-offset: 2px;
  outline-style: solid;
}

.c1 .fi-filter-input_input-element-container:focus-within > input:focus {
  outline: none;
}

.c1 .fi-filter-input_input-element-container:focus-within {
  position: relative;
}

.c1 .fi-filter-input_input-element-container:focus-within::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c1 .fi-filter-input_input {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  min-width: 245px;
  max-width: 100%;
  padding: 8px 16px;
  border: 1px solid hsl(202,7%,80%);
  border-radius: 2px;
  line-height: 1;
  background-color: hsl(0,0%,100%);
  min-width: 40px;
  width: 100%;
  min-height: 40px;
  padding-left: 10px;
  border-color: hsl(201,7%,58%);
}

.c1 .fi-filter-input_input::-webkit-input-placeholder {
  font-style: italic;
  color: hsl(201,7%,46%);
  opacity: 1;
}

.c1 .fi-filter-input_input::-moz-placeholder {
  font-style: italic;
  color: hsl(201,7%,46%);
  opacity: 1;
}

.c1 .fi-filter-input_input:-ms-input-placeholder {
  font-style: italic;
  color: hsl(201,7%,46%);
  opacity: 1;
}

.c1 .fi-filter-input_input::placeholder {
  font-style: italic;
  color: hsl(201,7%,46%);
  opacity: 1;
}

.c1.fi-filter-input--disabled .fi-text-input_input {
  color: hsl(202,7%,67%);
  background-color: hsl(202,7%,97%);
}

<div
  class="c0 fi-filter-input c1"
>
  <label
    class="c0 c2 fi-label-text"
    for="1"
  >
    <span
      class="c3 fi-label-text_label-span"
    >
      Label
    </span>
  </label>
  <div
    class="c0 fi-filter-input_input-element-container"
  >
    <input
      aria-autocomplete="list"
      auto-capitalize="false"
      autocomplete="off"
      class="c4 fi-filter-input_input"
      id="1"
      spellcheck="false"
      type="text"
    />
  </div>
</div>
`;
